std/experimental/allocator/package.d

```diff
@@ -763,10 +772,18 @@ if (isInputRange!R && !isInfinite!R)
             alloc.deallocate(m);
         }
 
-        for (; !range.empty; range.popFront, ++i)
+        import std.conv : emplace;
+        static if (isNarrowString!R || isRandomAccessRange!R)
```

Comment:
we need the special case for narrow strings & as we already have the special case, using it for arrays will be faster too (see e.g. #4265)